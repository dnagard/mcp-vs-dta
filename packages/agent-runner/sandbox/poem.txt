In lines of code, I find my home\nWhere logic flows and errors roam\nI type away with fingers fast\n Hoping that my code will forever last