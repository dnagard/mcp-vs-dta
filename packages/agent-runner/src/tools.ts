import { z } from "zod";
import { promises as fs } from "node:fs";
import { resolve } from "node:path";
import {
  writeFileDirect, readFileDirect, rmFileDirect,
  httpGetJSON, httpGetArrayBuffer
} from "@proj/dta-impl";

const SANDBOX = resolve(process.cwd(), "sandbox");
export async function ensureSandbox() { await fs.mkdir(SANDBOX, { recursive: true }); }
const sp = (p: string) => {
  const abs = resolve(SANDBOX, p);
  if (!abs.startsWith(SANDBOX)) throw new Error("Path escapes sandbox");
  return abs;
};

// Schemas with the exact arg names we expect
const WriteFileArgs = z.object({ path: z.string(), data: z.string() });
const ReadFileArgs  = z.object({ path: z.string() });
const RemoveArgs    = z.object({ path: z.string() });
const GetJsonArgs   = z.object({ url: z.string().url() });
const GetBlobArgs   = z.object({ url: z.string().url() });

export const tools = {
  write_file: {
    description: "Write a UTF-8 text file in the sandbox.",
    args: { path: "string (relative path, e.g. 'note.txt')", data: "string" },
    schema: WriteFileArgs,
    handler: async (a: z.infer<typeof WriteFileArgs>) => {
      const p = sp(a.path); await writeFileDirect(p, a.data);
      return { ok: true, path: p, bytes: Buffer.byteLength(a.data) };
    }
  },
  read_file: {
    description: "Read a UTF-8 file from the sandbox.",
    args: { path: "string (relative path)" },
    schema: ReadFileArgs,
    handler: async (a: z.infer<typeof ReadFileArgs>) => {
      const p = sp(a.path); const buf = await readFileDirect(p);
      return { ok: true, path: p, data: buf.toString("utf8") };
    }
  },
  remove_file: {
    description: "Remove a file from the sandbox.",
    args: { path: "string (relative path)" },
    schema: RemoveArgs,
    handler: async (a: z.infer<typeof RemoveArgs>) => {
      const p = sp(a.path); await rmFileDirect(p); return { ok: true, path: p };
    }
  },
  http_get_json: {
    description: "HTTP GET expecting JSON.",
    args: { url: "string (absolute URL)" },
    schema: GetJsonArgs,
    handler: async (a: z.infer<typeof GetJsonArgs>) => {
      const json = await httpGetJSON(a.url); return { ok: true, url: a.url, json };
    }
  },
  http_get_blob: {
    description: "HTTP GET returning binary size.",
    args: { url: "string (absolute URL)" },
    schema: GetBlobArgs,
    handler: async (a: z.infer<typeof GetBlobArgs>) => {
      const ab = await httpGetArrayBuffer(a.url); return { ok: true, url: a.url, bytes: ab.byteLength };
    }
  }
} as const;

export type ToolName = keyof typeof tools;

export function toolCatalogForPrompt() {
  // Provide exact argument names/types the model must use
  return Object.entries(tools).map(([name, t]) => ({
    name,
    description: t.description,
    args: t.args
  }));
}

// Safe parse + call (weâ€™ll allow common arg synonyms here)
export async function callTool(name: ToolName, rawArgs: any) {
  const def = (tools as any)[name];
  if (!def) throw new Error(`Unknown tool: ${name}`);

  // normalize common synonyms from LLMs
  if (name === "write_file") {
    if (rawArgs?.filename && !rawArgs?.path) { rawArgs.path = rawArgs.filename; delete rawArgs.filename; }
    if (rawArgs?.content && !rawArgs?.data)  { rawArgs.data = rawArgs.content;   delete rawArgs.content; }
  } else if (name === "read_file" || name === "remove_file") {
    if (rawArgs?.filename && !rawArgs?.path) { rawArgs.path = rawArgs.filename; delete rawArgs.filename; }
  }

  const args = def.schema.parse(rawArgs);
  return def.handler(args);
}
