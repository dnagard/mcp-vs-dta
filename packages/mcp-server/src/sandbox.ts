import { promises as fs } from "node:fs";
import { resolve, sep } from "node:path";

export class SandboxPathError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SandboxPathError";
  }
}

export function getSandboxRoot(): string {
  const rootEnv = process.env.MCP_SANDBOX_ROOT;
  const base =
    rootEnv && rootEnv.trim().length > 0
      ? rootEnv
      : resolve(process.cwd(), "sandbox");
  return resolve(base);
}

export async function ensureSandboxRoot(root = getSandboxRoot()) {
  await fs.mkdir(root, { recursive: true });
  return root;
}

export function resolveSandbox(root: string, relativePath: string): string {
  const absRoot = resolve(root);
  const abs = resolve(absRoot, relativePath);
  const normalizedRoot = absRoot.endsWith(sep) ? absRoot : absRoot + sep;
  if (abs !== absRoot && !abs.startsWith(normalizedRoot)) {
    throw new SandboxPathError(`Path escapes sandbox: ${relativePath}`);
  }
  return abs;
}
