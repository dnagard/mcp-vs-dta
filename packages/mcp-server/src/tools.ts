import { z } from "zod";
import {
  writeFileDirect,
  readFileDirect,
  rmFileDirect,
  httpGetJSON,
  httpGetArrayBuffer,
} from "@proj/dta-impl";
import { resolveSandbox } from "./sandbox.js";

export class ToolInputError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ToolInputError";
  }
}

export const toolSchemas = {
  write_file: {
    name: "write_file",
    description: "Write a UTF-8 text file in the sandbox.",
    input_schema: {
      type: "object",
      properties: {
        path: {
          type: "string",
          description: "Relative path under sandbox, e.g. 'note.txt'",
        },
        data: {
          type: "string",
          description: "UTF-8 file contents",
        },
      },
      required: ["path", "data"],
      additionalProperties: false,
    },
    schema: z.object({
      path: z.string(),
      data: z.string(),
    }),
  },
  read_file: {
    name: "read_file",
    description: "Read a UTF-8 file from the sandbox.",
    input_schema: {
      type: "object",
      properties: {
        path: {
          type: "string",
          description: "Relative path under sandbox",
        },
      },
      required: ["path"],
      additionalProperties: false,
    },
    schema: z.object({
      path: z.string(),
    }),
  },
  remove_file: {
    name: "remove_file",
    description: "Remove a file from the sandbox.",
    input_schema: {
      type: "object",
      properties: {
        path: {
          type: "string",
          description: "Relative path under sandbox",
        },
      },
      required: ["path"],
      additionalProperties: false,
    },
    schema: z.object({
      path: z.string(),
    }),
  },
  http_get_json: {
    name: "http_get_json",
    description: "HTTP GET expecting JSON.",
    input_schema: {
      type: "object",
      properties: {
        url: {
          type: "string",
          format: "uri",
          description: "Absolute URL",
        },
      },
      required: ["url"],
      additionalProperties: false,
    },
    schema: z.object({
      url: z.string().url(),
    }),
  },
  http_get_blob: {
    name: "http_get_blob",
    description: "HTTP GET returning binary size.",
    input_schema: {
      type: "object",
      properties: {
        url: {
          type: "string",
          format: "uri",
          description: "Absolute URL",
        },
      },
      required: ["url"],
      additionalProperties: false,
    },
    schema: z.object({
      url: z.string().url(),
    }),
  },
} as const;

export type ToolName = keyof typeof toolSchemas;

type InferSchema<T extends ToolName> = z.infer<(typeof toolSchemas)[T]["schema"]>;

export function listTools() {
  return Object.values(toolSchemas).map((tool) => ({
    name: tool.name,
    description: tool.description,
    input_schema: tool.input_schema,
  }));
}

export function validate<T extends ToolName>(name: T, args: unknown): InferSchema<T> {
  const def = toolSchemas[name];
  if (!def) {
    throw new Error(`Unknown tool: ${name}`);
  }
  return def.schema.parse(args);
}

type ToolContext = {
  sandboxRoot: string;
};

const handlers = {
  async write_file(args: InferSchema<"write_file">, ctx: ToolContext) {
    const path = resolveSandbox(ctx.sandboxRoot, args.path);
    await writeFileDirect(path, args.data);
    return {
      ok: true,
      path,
      bytes: Buffer.byteLength(args.data, "utf8"),
    };
  },
  async read_file(args: InferSchema<"read_file">, ctx: ToolContext) {
    const path = resolveSandbox(ctx.sandboxRoot, args.path);
    const buf = await readFileDirect(path);
    return {
      ok: true,
      path,
      data: buf.toString("utf8"),
    };
  },
  async remove_file(args: InferSchema<"remove_file">, ctx: ToolContext) {
    const path = resolveSandbox(ctx.sandboxRoot, args.path);
    await rmFileDirect(path);
    return {
      ok: true,
      path,
    };
  },
  async http_get_json(args: InferSchema<"http_get_json">, _ctx: ToolContext) {
    const url = new URL(args.url);
    if (!isHttpProtocol(url.protocol)) {
      throw new ToolInputError(`Unsupported URL protocol: ${url.protocol || "unknown"}`);
    }
    const json = await httpGetJSON(url.toString());
    return {
      ok: true,
      url: url.toString(),
      json,
    };
  },
  async http_get_blob(args: InferSchema<"http_get_blob">, _ctx: ToolContext) {
    const url = new URL(args.url);
    if (!isHttpProtocol(url.protocol)) {
      throw new ToolInputError(`Unsupported URL protocol: ${url.protocol || "unknown"}`);
    }
    const buffer = await httpGetArrayBuffer(url.toString());
    return {
      ok: true,
      url: url.toString(),
      bytes: buffer.byteLength,
    };
  },
} satisfies {
  [K in ToolName]: (args: InferSchema<K>, ctx: ToolContext) => Promise<any>;
};

export async function executeTool<T extends ToolName>(
  name: T,
  args: InferSchema<T>,
  ctx: ToolContext,
) {
  const handler = handlers[name] as (args: InferSchema<T>, ctx: ToolContext) => Promise<any>;
  return handler(args, ctx);
}

export function isToolName(name: string): name is ToolName {
  return Object.prototype.hasOwnProperty.call(toolSchemas, name);
}

function isHttpProtocol(protocol: string) {
  return protocol === "http:" || protocol === "https:";
}
