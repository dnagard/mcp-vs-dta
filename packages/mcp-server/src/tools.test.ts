import { randomUUID } from "node:crypto";
import { promises as fs } from "node:fs";
import { mkdtemp } from "node:fs/promises";
import http from "node:http";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { ZodError } from "zod";
import { listTools, validate, executeTool, ToolInputError } from "./tools.js";
import { resolveSandbox, SandboxPathError } from "./sandbox.js";

describe("tool catalog", () => {
  it("matches expected shape", () => {
    expect(listTools()).toMatchSnapshot();
  });
});

describe("filesystem tools", () => {
  let sandboxRoot: string;

  beforeEach(async () => {
    sandboxRoot = await mkdtemp(join(tmpdir(), "mcp-server-test-"));
  });

  afterEach(async () => {
    await fs.rm(sandboxRoot, { recursive: true, force: true });
  });

  it("writes, reads, and removes files", async () => {
    const path = `file-${randomUUID()}.txt`;
    await executeTool("write_file", { path, data: "hello" }, { sandboxRoot });
    const read = await executeTool("read_file", { path }, { sandboxRoot });
    expect(read).toMatchObject({ ok: true, data: "hello" });
    await executeTool("remove_file", { path }, { sandboxRoot });
    await expect(fs.access(resolveSandbox(sandboxRoot, path))).rejects.toThrow();
  });

  it("rejects sandbox escape attempts", async () => {
    await expect(
      executeTool("write_file", { path: "../../etc/passwd", data: "nope" }, { sandboxRoot })
    ).rejects.toBeInstanceOf(SandboxPathError);
  });

  it("returns IO errors for missing files", async () => {
    await expect(executeTool("read_file", { path: "missing.txt" }, { sandboxRoot })).rejects.toThrow();
  });
});

describe("http tools", () => {
  let server: http.Server;
  let urlBase: string;

  beforeEach(async () => {
    server = http.createServer((req, res) => {
      if (req.url?.startsWith("/json")) {
        res.writeHead(200, { "content-type": "application/json" });
        res.end(JSON.stringify({ ok: true }));
        return;
      }
      if (req.url?.startsWith("/blob")) {
        res.writeHead(200, { "content-type": "application/octet-stream" });
        res.end(Buffer.from([1, 2, 3, 4]));
        return;
      }
      res.writeHead(404).end();
    });
    await new Promise<void>((resolvePromise) => server.listen(0, resolvePromise));
    const address = server.address();
    if (typeof address === "object" && address) {
      urlBase = `http://127.0.0.1:${address.port}`;
    } else {
      throw new Error("Failed to start test server");
    }
  });

  afterEach(async () => {
    await new Promise<void>((resolvePromise, rejectPromise) =>
      server.close((err) => (err ? rejectPromise(err) : resolvePromise()))
    );
  });

  it("fetches JSON payloads", async () => {
    const result = await executeTool(
      "http_get_json",
      { url: `${urlBase}/json` },
      { sandboxRoot: tmpdir() }
    );
    expect(result).toMatchObject({ ok: true, json: { ok: true } });
  });

  it("fetches binary payload sizes", async () => {
    const result = await executeTool(
      "http_get_blob",
      { url: `${urlBase}/blob` },
      { sandboxRoot: tmpdir() }
    );
    expect(result).toMatchObject({ ok: true, bytes: 4 });
  });

  it("rejects unsupported schemes", async () => {
    await expect(
      executeTool("http_get_json", { url: "ftp://example.com" }, { sandboxRoot: tmpdir() })
    ).rejects.toBeInstanceOf(ToolInputError);
  });
});

describe("validation", () => {
  it("surfaces zod issues for missing data", () => {
    expect(() => validate("write_file", { path: "a" } as any)).toThrow(ZodError);
  });

  it("throws on unknown tools", () => {
    expect(() => validate("does_not_exist" as any, {})).toThrow(/Unknown tool/);
  });
});
