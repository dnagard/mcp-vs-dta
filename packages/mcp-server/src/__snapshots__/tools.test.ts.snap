// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tool catalog > matches expected shape 1`] = `
[
  {
    "description": "Write a UTF-8 text file in the sandbox.",
    "input_schema": {
      "additionalProperties": false,
      "properties": {
        "data": {
          "description": "UTF-8 file contents",
          "type": "string",
        },
        "path": {
          "description": "Relative path under sandbox, e.g. 'note.txt'",
          "type": "string",
        },
      },
      "required": [
        "path",
        "data",
      ],
      "type": "object",
    },
    "name": "write_file",
  },
  {
    "description": "Read a UTF-8 file from the sandbox.",
    "input_schema": {
      "additionalProperties": false,
      "properties": {
        "path": {
          "description": "Relative path under sandbox",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "type": "object",
    },
    "name": "read_file",
  },
  {
    "description": "Remove a file from the sandbox.",
    "input_schema": {
      "additionalProperties": false,
      "properties": {
        "path": {
          "description": "Relative path under sandbox",
          "type": "string",
        },
      },
      "required": [
        "path",
      ],
      "type": "object",
    },
    "name": "remove_file",
  },
  {
    "description": "HTTP GET expecting JSON.",
    "input_schema": {
      "additionalProperties": false,
      "properties": {
        "url": {
          "description": "Absolute URL",
          "format": "uri",
          "type": "string",
        },
      },
      "required": [
        "url",
      ],
      "type": "object",
    },
    "name": "http_get_json",
  },
  {
    "description": "HTTP GET returning binary size.",
    "input_schema": {
      "additionalProperties": false,
      "properties": {
        "url": {
          "description": "Absolute URL",
          "format": "uri",
          "type": "string",
        },
      },
      "required": [
        "url",
      ],
      "type": "object",
    },
    "name": "http_get_blob",
  },
]
`;
